### ISSUE TITLE
Data serialization of types is pluggable to enable sharing of data 

### ISSUE THREAD

**Issue by artursouza** [dapr/dapr#2905]
## Current State

Today, there is a difference in how SDKs handle serialization of application domain objects.

* .Net SDK: uses JSON serialization
* Java SDK: recommends customers to plugin their own serializer, using default only for examples and tests.
* Python SDK: does not provide any serialization, input and output is done via bytes directly
* PHP SDK: uses JSON serialization

## Goal

The goal of this proposal is to agree on a path forward to bring consistency across SDKs regarding serialization of application domain objects.

### 1. Adopt JSON as the only serialization protocol

TL;DR: Like in the .Net and PHP SDK.

* Simple to use and other protocols can be used but would be serialized into valid JSON (string or base64 encoded byte[])
* Must handle "infinite" different classes.
* SDKs have control over compatibility
* Higher chances of breaking changes since serialization is part of the contract
* Compatibility issues in serialization must consider persisted data
* Must version serialization to avoid breaking changes

### 2. Make serialization a pluggable component

TL;DR: Like in the Java SDK.

* Offers a mechanism to use any serializer with the SDK (JSON, XML, etc) - each SDK finds the best way in how this is achieved
* SDK does not interfere with serialization of persisted data

### 3. Only accept byte array

TL;DR: Like in the Python SDK.

* Simple implementation
* App must handle serialization before/after invoking Dapr APIs.
* SDK does not interfere with serialization of persisted data

> Note: Dapr domain objects is not under discussion here because the serialization/deserialization of these objects are determined by the Dapr API.

> Note 2: We will update this description as discussion evolves.

> Note 3: We can also collect votes after discussions have settled.

**Comment by mukundansundar** [comment]
My option would be 2.
Provides good flexibility. Puts onus on app developer to provide the serialization logic for application objects while still allowing a simple way for integrating that with the SDK. 

Cons for option 1: Version changes of the serialization library must take into account the already persisted data and handle breaking changes for all objects (not only Dapr specific). 

Cons for option 3: For hand crafted SDKs, losing the native language experience of passing in objects to method calls and only passing in bytes is not good IMO.

**Comment by wcs1only** [comment]
I'm definitely in favor of 2 or 3. Plenty of languages offer richer serialization and features beyond JSON (python pickle, for example). I think it is presumptuous to assume/force JSON. We could also define 3 as a special case of 2, if we really wanted.

**Comment by xiazuojie** [comment]
I'd like to propose a strategy as discussed [here](https://github.com/dapr/java-sdk/issues/500).

# Serialization negotiation between client and server based on metadata.
At the moment, serialization type is pre-agreed between client and server. So both sides have to agree on a single serialization type. Due to this, richer serialization and features can not be utilized.

If we add fields to [request and response objects](https://github.com/dapr/dapr/blob/master/dapr/proto/runtime/v1/appcallback.proto) for all Dapr APIs.
* data-content-type: the media type of the data, e.g. application/json, text/plain, application/octet-stream. Should be specified in both request and response objects.
* data-accept: optional. The content-type the client supports. The server should prefer do serialize data supported by the client.

As you can see, it's the same idea of [Content_negotiation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation) used by HTTP.

**Comment by artursouza** [comment]
@xiazuojie JSON does not offer cross language compatibility. Many classes (DateTime, for example), are serialized in a specific way on each app. This issue is about defining what is the common denominator for serializing cross SDKs and I do agree that the metadata regarding content-type is key but is not enough. Each app has their own way of serializing specific data types and the compatibility cross language is not something Dapr can resolve - each app owns that.

**Comment by rynowak** [comment]
Commenting here because @artursouza asked me to -- I'm mostly going to disagree with everything that's being discussed here. I'll do my best to keep it constructive, so please interpret my comments with best intentions. 

I'm honestly a little confused by this topic being brought up. I'm not sure why we'd consider making any drastic changes or what's to be gained by a broad *all Dapr* discussion. It feels to me like the whole Dapr ecosystem is already in a very stable place re: serialization and interop already.

---

I think my issue is that I just don't fundamentally agree with the goal as stated, and so I won't agree with the proposals 😆 

> The goal of this proposal is to agree on a path forward to bring consistency across SDKs regarding serialization of application domain objects.

If you take the *broad view* of the term consistency - then I don't actually care about *consistency* as a goal, and I think it's the wrong thing to aim for. 

If I'm writing a Go application, I expect the Go Dapr SDK to feel like it was designed for me to in a Go application. I would be unhappy as a user if the Go SDK for Dapr was designed like a Java library or a C# library. Furthermore if I need to send requests to/from a C# application from my Go application, the libraries should not make it harder.

What I actually care about is:

- **idiomaticicity:** (not sure if that's a word) the library I'm using feels natural to me as a developer in my chosen language
- **interoperability:** can code written in one language interact with service written in another

----

So how do we do this...?

I think we already do a good job writing idiomatic libraries. All we have to do is keep doing that. 

However, I think we have to break interoperability into two categories. We can't really say that Dapr **guarantees** that you will write interoperable, polyglot applications, we can only say that Dapr **helps you** write interoperable polyglot applications.

- Things that Dapr controls or influences
- Things that are out of scope for Dapr

In my mind what we control is the set of experiences we enable through the Dapr protocol(s) and building blocks, and well as the set of experiences we make easy by making them first class in the SDKs.

So for instance it hurts interoperability if we push users to use Protobuf from Python but make it really hard to do from Go, etc. It hurts interoperability if C# supports sending arbitrary binary content via pub/sub but PHP doesn't make it possible to consume. 

I think we have to acknowledge, and then let go of some of the things that are out of scope for Dapr. JSON as an ecosystem has well-understood limitations, and that doesn't stop it being a useful and interoperable format for millions of developers. We can acknowledge that there are inconsistencies in how languages/libraries work with JSON, and we can acknowledge that JSON doesn't attempt to solve versioning. Acknowledging those limitations and moving on puts us on equal or better footing with every other technology out there 😆.

I think some of the ideas already in progress or already implemented in our SDKs, like giving users a choice over which serializer to use, or exposing that serializers settings for customization are the best solutions we can pursue. 

I think it would be really **counterproductive** to make a custom serialization format for Dapr, or to make a sub-dialect within Dapr to try and solve a problem like versioning. This raises so many questions that now need answers, like how does non-Daprized code interop with us. 

---

To these ends, here's what I'd propose we agree on:

- Every SDK **must** expose 'raw' functionality for each building block that lets users send/consume any format (bytes)
- Places where the Dapr protocol includes arbitrary content, but doesn't include content type should have it added
- Our documentation and samples for each SDK should cover JSON as an example and a default recommendation
- Our documentation and samples for each SDK should include an example of a non-JSON format
- The rest is up to the SDK to do whatever makes sense for that language ecosystem. If that means a pluggable serializer abstraction, then do that. it's just a matter of what's idiomatic for users - for instance a *pluggable serializer* as stated here is a very object-oriented-oriented solution. We should not be one-size-fits-all at the *code* level.

How is this different from the proposals here and the current state of the art:

I don't think we consistently provide a good experience for 'raw' content and we need to. This is a very pragmatic layering concern, and it unblocks users taking full control if they need to. We should make this a priority.

I don't propose that we limit or stipulate what SDKs are allowed to provide as a built-in feature. For instance Go does not provide JSON or any other format as a built-in feature, and that's super idiomatic for Go. 

I propose that we put the onus on the SDK owner to plan for multi-format support in a way that makes sense for that ecosystem. I think JSON is the best choice for the majority of people right now. Maybe it's going to be out of date in 3 years.

**Comment by artursouza** [comment]
Thanks Ryan for your deep feedback on many fronts here. Let me try to understand and maybe reply to some points here without deviating from this issue.

> Commenting here because @artursouza asked me to -- I'm mostly going to disagree with everything that's being discussed here. I'll do my best to keep it constructive, so please interpret my comments with best intentions.
> 
> I'm honestly a little confused by this topic being brought up. I'm not sure why we'd consider making any drastic changes or what's to be gained by a broad _all Dapr_ discussion. It feels to me like the whole Dapr ecosystem is already in a very stable place re: serialization and interop already.
This is not really proposing drastic changes. I agree that Dapr is definitely in a good shape right now regarding serialization and interop - it does not mean nothing should change and we should freeze where we are. For example, your mention "Places where the Dapr protocol includes arbitrary content, but doesn't include content type should have it added", which I agree and is a significant (not necessarily drastic) change.

One point that I fundamentally disagree is on "what's to be gained by a broad _all Dapr_ discussion". That is a dangerous path IMO, where maintainers work in isolation and avoid discussions. I actually believe in the opposite: constant communication among maintainers and the community in issues that cross cut Dapr - even if it means to discuss if it should cross cut (which might be what we discuss here too).
> 
> I think my issue is that I just don't fundamentally agree with the goal as stated, and so I won't agree with the proposals 😆
> 
> > The goal of this proposal is to agree on a path forward to bring consistency across SDKs regarding serialization of application domain objects.
> 
> If you take the _broad view_ of the term consistency - then I don't actually care about _consistency_ as a goal, and I think it's the wrong thing to aim for.
I have a different view on this, but thanks for voicing your opinion.
> 
> If I'm writing a Go application, I expect the Go Dapr SDK to feel like it was designed for me to in a Go application. I would be unhappy as a user if the Go SDK for Dapr was designed like a Java library or a C# library. Furthermore if I need to send requests to/from a C# application from my Go application, the libraries should not make it harder.
I agree 100% with this statement. This has been the way we have designed the SDKs so far and we should continue to do so.
> 
> What I actually care about is:
> 
> * **idiomaticicity:** (not sure if that's a word) the library I'm using feels natural to me as a developer in my chosen language
> * **interoperability:** can code written in one language interact with service written in another
> 
> So how do we do this...?
> 
> I think we already do a good job writing idiomatic libraries. All we have to do is keep doing that.
I agree with the statement above, 100%.
> 
> However, I think we have to break interoperability into two categories. We can't really say that Dapr **guarantees** that you will write interoperable, polyglot applications, we can only say that Dapr **helps you** write interoperable polyglot applications.
Great point. Agreed.
> 
> * Things that Dapr controls or influences
> * Things that are out of scope for Dapr
> 
> In my mind what we control is the set of experiences we enable through the Dapr protocol(s) and building blocks, and well as the set of experiences we make easy by making them first class in the SDKs.
Agreed.
> 
> So for instance it hurts interoperability if we push users to use Protobuf from Python but make it really hard to do from Go, etc. It hurts interoperability if C# supports sending arbitrary binary content via pub/sub but PHP doesn't make it possible to consume.
> 
> I think we have to acknowledge, and then let go of some of the things that are out of scope for Dapr. JSON as an ecosystem has well-understood limitations, and that doesn't stop it being a useful and interoperable format for millions of developers. We can acknowledge that there are inconsistencies in how languages/libraries work with JSON, and we can acknowledge that JSON doesn't attempt to solve versioning. Acknowledging those limitations and moving on puts us on equal or better footing with every other technology out there 😆.
I agree with this paragraph, 100%. Great point.
> 
> I think some of the ideas already in progress or already implemented in our SDKs, like giving users a choice over which serializer to use, or exposing that serializers settings for customization are the best solutions we can pursue.
This issue is about bringing consistency to this experience.
> 
> I think it would be really **counterproductive** to make a custom serialization format for Dapr, or to make a sub-dialect within Dapr to try and solve a problem like versioning. This raises so many questions that now need answers, like how does non-Daprized code interop with us.
This issue is **not** proposing "custom serialization format for Dapr". I am happy to edit the description of the issue if you can point out where this is mentioned.
> 
> To these ends, here's what I'd propose we agree on:
> 
> * Every SDK **must** expose 'raw' functionality for each building block that lets users send/consume any format (bytes)
Agree.
> * Places where the Dapr protocol includes arbitrary content, but doesn't include content type should have it added
Agree.
> * Our documentation and samples for each SDK should cover JSON as an example and a default recommendation
Partially agree, I am not so sure about "default recommendation".
> * Our documentation and samples for each SDK should include an example of a non-JSON format
Agree.
> * The rest is up to the SDK to do whatever makes sense for that language ecosystem. If that means a pluggable serializer abstraction, then do that. it's just a matter of what's idiomatic for users - for instance a _pluggable serializer_ as stated here is a very object-oriented-oriented solution. We should not be one-size-fits-all at the _code_ level.
Partially agree. I don't think that "pluggable serializer" means "object-oriented-oriented" solutions - there are many ways languages allow extensibility beyond OO. We have to be careful when interpreting "up to the SDK to do whatever makes sense for that language ecosystem" because it can mean an extreme point where there is total lack on consistency - I remember that we had to rename things to be consistent across SDKs, for example.
> 
> How is this different from the proposals here and the current state of the art:
> 
> I don't think we consistently provide a good experience for 'raw' content and we need to. This is a very pragmatic layering concern, and it unblocks users taking full control if they need to. We should make this a priority.
I agree. This issue is a step in that direction.
> 
> I don't propose that we limit or stipulate what SDKs are allowed to provide as a built-in feature. For instance Go does not provide JSON or any other format as a built-in feature, and that's super idiomatic for Go.
I agree that we should not "impose" things but there is some level of consistency. Again, I just would like to avoid extremes here where all SDKs are "forcefully the same" or "free for all".
> 
> I propose that we put the onus on the SDK owner to plan for multi-format support in a way that makes sense for that ecosystem. I think JSON is the best choice for the majority of people right now. Maybe it's going to be out of date in 3 years. If we end up growing the set of things that are _first class_ over time, that's actually best - that way we're being driven by what the community asking for.
This issue is not about designing for each SDK but about a high level agreement on extensibility of serialization in the SDKs.

**Comment by rynowak** [comment]
> Partially agree. I don't think that "pluggable serializer" means "object-oriented-oriented" solutions - there are many ways languages allow extensibility beyond OO. We have to be careful when interpreting "up to the SDK to do whatever makes sense for that language ecosystem" because it can mean an extreme point where there is total lack on consistency - I remember that we had to rename things to be consistent across SDKs, for example.

I think I might be potentially nitpicking again - but I don't think there's value in **consistency** without additional surrounding context. I want to work though exactly what you are saying needs to be consistent, and what the value of that is in order to get on the same page.

I think I mostly agree if we modify one of your statements in option 2:

**From**

> Offers an interface for easy integration of serializers and the SDK

**To**

> Offers a mechanism to use any serializer with the SDK

I think from a user POV the resulting value and set of scenarios that work is the same but there are fewer assumptions in how its being stated. 

--- 

Something else I forgot to mention earlier:

> SDK does not interfere with serialization of persisted data

In my experience this this is absolutely a great choice. I've worked on projects in the past where we injected opinions of our own and it becomes really tiring to support users.

**Comment by artursouza** [comment]
> > Partially agree. I don't think that "pluggable serializer" means "object-oriented-oriented" solutions - there are many ways languages allow extensibility beyond OO. We have to be careful when interpreting "up to the SDK to do whatever makes sense for that language ecosystem" because it can mean an extreme point where there is total lack on consistency - I remember that we had to rename things to be consistent across SDKs, for example.
> 
> I think I might be potentially nitpicking again - but I don't think there's value in **consistency** without additional surrounding context. I want to work though exactly what you are saying needs to be consistent, and what the value of that is in order to get on the same page.
> 
> I think I mostly agree if we modify one of your statements in option 2:
> 
> **From**
> 
> > Offers an interface for easy integration of serializers and the SDK
> 
> **To**
> 
> > Offers a mechanism to use any serializer with the SDK
> 
> I think from a user POV the resulting value and set of scenarios that work is the same but there are fewer assumptions in how its being stated.
> 
> Something else I forgot to mention earlier:
> 
> > SDK does not interfere with serialization of persisted data
> 
> In my experience this this is absolutely a great choice. I've worked on projects in the past where we injected opinions of our own and it becomes really tiring to support users.

Changed that statement like you proposed, with the addition on wording that SDKs define *how* it is designed. At this point, I assume we are on the same page.

**Comment by xiazuojie** [comment]
> My idea is to keep serialization as pluggable component and add the metadata about content-type, like you suggest. But **do not** try to dictate how apps serialize, specially as the number of languages we support increase.

I agree that serialization should not be part of spec. I suggest that default value of `data-content-type` to be `application/json` only in order to make this a non-breaking change. 
I'd like to PR for pluggable serializer for Java SDK if `data-content-type` and `data-accept` is introduced in [dapr.proto](https://github.com/dapr/dapr/blob/master/dapr/proto/runtime/v1/dapr.proto)

**Comment by artursouza** [comment]
> > My idea is to keep serialization as pluggable component and add the metadata about content-type, like you suggest. But **do not** try to dictate how apps serialize, specially as the number of languages we support increase.
> 
> I agree that serialization type should not be part of spec. I suggest that default value of `data-content-type` to be `application/json` in order to make this a non-breaking change.
> I'd like to PR for pluggable serializer for Java SDK if `data-content-type` and `data-accept` is introduced in [dapr.proto](https://github.com/dapr/dapr/blob/master/dapr/proto/runtime/v1/dapr.proto)

@xiazuojie We already have a pluggable serializer in the Java SDK. There is a new issue discussing how to expose multiple serializer implementations in the Java SDK: https://github.com/dapr/java-sdk/issues/509

I actually wonder if this work needs to wait for `data-content-type` and `data-accept`.

**Comment by xiazuojie** [comment]
> > > My idea is to keep serialization as pluggable component and add the metadata about content-type, like you suggest. But **do not** try to dictate how apps serialize, specially as the number of languages we support increase.
> > 
> > 
> > I agree that serialization type should not be part of spec. I suggest that default value of `data-content-type` to be `application/json` in order to make this a non-breaking change.
> > I'd like to PR for pluggable serializer for Java SDK if `data-content-type` and `data-accept` is introduced in [dapr.proto](https://github.com/dapr/dapr/blob/master/dapr/proto/runtime/v1/dapr.proto)
> 
> @xiazuojie We already have a pluggable serializer in the Java SDK. There is a new issue discussing how to expose multiple serializer implementations in the Java SDK: [dapr/java-sdk#509](https://github.com/dapr/java-sdk/issues/509)
> 
> I actually wonder if this work needs to wait for `data-content-type` and `data-accept`.

Yes. Java SDK already has pluggable serializer. But "pluggable" means serialization strategy can be user-specified. It is not the same as "negotiated". To support serialization-negotiation, The DaprObjectSerializer interface needs be updated to take account of `data-content-type` and `data-accept` parameters of requests and responses.

**Comment by vikas2k14** [comment]
> > > > My idea is to keep serialization as pluggable component and add the metadata about content-type, like you suggest. But **do not** try to dictate how apps serialize, specially as the number of languages we support increase.
> > > 
> > > 
> > > I agree that serialization type should not be part of spec. I suggest that default value of `data-content-type` to be `application/json` in order to make this a non-breaking change.
> > > I'd like to PR for pluggable serializer for Java SDK if `data-content-type` and `data-accept` is introduced in [dapr.proto](https://github.com/dapr/dapr/blob/master/dapr/proto/runtime/v1/dapr.proto)
> > 
> > 
> > @xiazuojie We already have a pluggable serializer in the Java SDK. There is a new issue discussing how to expose multiple serializer implementations in the Java SDK: [dapr/java-sdk#509](https://github.com/dapr/java-sdk/issues/509)
> > I actually wonder if this work needs to wait for `data-content-type` and `data-accept`.
> 
> Yes. Java SDK already has pluggable serializer. But "pluggable" means serialization strategy can be user-specified. It is not the same as "negotiated". To support serialization-negotiation, The DaprObjectSerializer interface needs to be updated to take account of `data-content-type` and `data-accept` parameters of requests and responses.

👍  This is one of the hot requirements for us.

In fact, we wanted to write a generic serializer/deserializer but since DaprObjectSerializer doesn't take any parameter for content type and data-accept and always returns application/json we couldn't do it.

It would be really helpful to have this included.

Thanks.

cc: @rahulopengts @mukundansundar @akhila

**Comment by olitomlinson** [comment]
I would bias making a decision towards 

- What provides the simplest out-of-the-box interoperability, when serializing/deserializing objects across different languages.
- What gives the simplest out-of-the-box experience in the least lines of boiler plate.

**Example**

I don't know how possible this is, but given that a significant part of Daprs mission is ease of interoperability between micro-services (and there for cross-languages), if I use .NET Dapr SDK to serialize an object and persist into a state store, I expect that the default Java Dapr SDK configuration would be able to successfully deserialize that Object when reading from the State Store . And vice versa.
